cmd/zpool/zpool_vdev.c

```diff
@@ -1463,6 +1463,8 @@ construct_spec(nvlist_t *props, int argc, char **argv)
 					    gettext("invalid vdev "
 					    "specification: 'spare' can be "
 					    "specified only once\n"));
+					free(top);
+					free(spares);
 					return (NULL);
```

Comment:
Often a convenient way to handle messy cleanup like this is place all the cleanup at the end of the function.  Then for each error case you can set the return value to NULL and use a `goto` to jump to an `out` label which handles the cleanup.  In this case it looks like it would make sense to add an `out` label [here](https://github.com/zfsonlinux/zfs/pull/5143/commits/d658982d38352347642f3e9388fe8a8239782632#diff-ea3469b2eb70e3b5a05c22cf1a8ed06bR1642) and the various NULL returns could be updated to just.  This way you're always guaranteed these local allocations are freed.

```
nvroot = NULL;
goto out;
```